File: 2024/Python/03.py
import re

data = "".join(open("../inputs/03.txt").read().splitlines())

sum_one = sum(int(a) * int(b) for a, b in re.findall(r"mul\((\d+),(\d+)\)", data))
print(sum_one)

do = True
sum = 0

for match in re.finditer(r"(?:do\(\)|don\'t\(\))|mul\((\d+),(\d+)\)", data):
    if match.group(0) == "do()":
        do = True
    elif match.group(0) == "don't()":
        do = False
    else:
        sum += int(match.group(1)) * int(match.group(2)) if do else 0

print(sum)

File: 2024/Deno/03.ts
const data = (await Deno.readTextFile("../inputs/03.txt")).split("\n").join("");

const sum1 = [...data.matchAll(/mul\((\d+),(\d+)\)/g)]
    .reduce((sum, [_, a, b]) => sum + (Number(a) * Number(b)), 0);

console.log(sum1);

let enabled = true;

const sum2 = [...data.matchAll(/mul\((\d+),(\d+)\)|do\(\)|don't\(\)/g)]
    .reduce((sum, [m, a, b]) => {
        if (m === "do()") return enabled = true, sum;
        if (m === "don't()") return enabled = false, sum;
        return enabled ? sum + (Number(a) + Number(b)) : sum;
    }, 0);

console.log(sum2);
